auth/src/main/java/com/firebase/ui/auth/util/data/SingleLiveEvent.java

```diff
@@ -0,0 +1,99 @@
+package com.firebase.ui.auth.util.data;
+
+import android.arch.lifecycle.LifecycleOwner;
+import android.arch.lifecycle.MutableLiveData;
+import android.arch.lifecycle.Observer;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A LiveData that only notifies observers when new data arrives, used for events like navigation
+ * and SnackBar messages.
+ * <p>
+ * This avoids a common problem with events: on configuration change (like rotation) an update can
+ * be emitted if the observer is active.
+ * <p>
+ * This LiveData supports multiple observers per class definition, but works on a stack model. Only
+ * the newest observers get updates.
+ */
+public class SingleLiveEvent<T> extends MutableLiveData<T> {
```

Comment:
Ok I see what you are doing here but it took me a long time to sort it out so I'd suggest the following refactor:

  * Go back to the implementation from the sample that you linked to.  It is simple: no repeat updates.
  * Drop `EventFilterObserver`, I don't think we need it.
  * Make a class called like `SingleLiveEventRegistry<T>` or similar that keeps a map of `String --> SingleLiveEvent`
  * Rather than use classes for identification, have observers get this livedata by giving a `String` "TAG" to the registry.

So in your observer class you'd have something like:

```java
register.getLiveEvent("MyTag").observe(...)
```

I think that will simplify this a lot.   Correct me if I have missed the point of all this.

Reply:
@samtstern That would work, but I don't see how it will make things simpler since we'd just be moving the logic into another class (figuring out which generated live datas need to receive updates and keeping track of the real events etc...)